### DISCUSSION TITLE
Running didjvu without installing

### DISCUSSION THREAD

**Discussion by jsbien** [FriedrichFroebel/didjvu#24] [Q&A]
`Alternatively, you can use didjvu without installing it, straight out of an unpacked source tarball or a VCS checkout.`
Is this statement still true? If so, which file is to executed?

**Comment by FriedrichFroebel** [comment]
Didn't you already resolve this in #22?

And yes, this still holds true. You have to run the complete `didjvu` module - due to `didjvu/__main__.py`, this will run directly. Just use `python -m didjvu --help` from the repository root after installing the dependencies.

**Comment by jsbien** [comment]
> Didn't you already resolve this in #22?

If I remember well, I just decided to install the program after creating the virtual environment.

> 
> And yes, this still holds true. You have to run the complete `didjvu` module - due to `didjvu/__main__.py`, this will run directly. Just use `python -m didjvu --help` from the repository root after installing the dependencies.

A partial success: I get `usage: didjvu.__main__ [-h] [--version] {separate,encode,bundle} ...` but also
`didjvu.__main__: error: argument -h/--help: ignored explicit argument 'elp'`

Moreover I still don't know which file I should run from Thonny to be able to step into the proper code :-(

**Reply by FriedrichFroebel** [reply]
> If I remember well, I just decided to install the program after creating the virtual environment.

Running `pip uninstall didjvu` will uninstall *didjvu*, but keep the dependencies installed if you really want to work from source.

> A partial success: I get usage: `didjvu.__main__ [-h] [--version] {separate,encode,bundle} ...` but also
`didjvu.__main__: error: argument -h/--help: ignored explicit argument 'elp'`

It is either `-h` or `--help`, not `-help`.

> Moreover I still don't know which file I should run from Thonny to be able to step into the proper code :-(

The actual entry point is in `didjvu/__main__.py` - the installed `bin/didjvu` just points to the `didjvu.__main__.main` function. Nevertheless, you have to make sure that your `sys.path` is set correctly, as otherwise `__main__.py` might fail to run. Thus I would recommend to either go with the `bin/didjvu` file or with a custom `didjvu.py` file in the root directory of the repository:

```python3
from didjvu.__main__ import main

main()
```

Running `python didjvu.py --help` should give you the same results as before.

**Reply by jsbien** [reply]
Yes, this works. 
However when I try to debug didjvu.py in Thonny, it still does not allow me to step into main(). As this is a Thonny specific problem, I can ask on the Thonny mailing list.
On the other hand I don't understand the code of `__main__.py` :-( What is `didjvu.didjvu` in `from didjvu.didjvu import Main`? How the control is passed to the code proper?

**Reply by FriedrichFroebel** [reply]
`from didjvu.didjvu import Main` imports the internal main entry point https://github.com/FriedrichFroebel/didjvu/blob/5adbf8ba4402d57b805049e10f1f5c7a1f081caf/didjvu/didjvu.py#L214 whose `__init__` method starts the argument parsing and in https://github.com/FriedrichFroebel/didjvu/blob/5adbf8ba4402d57b805049e10f1f5c7a1f081caf/didjvu/cli.py#L340 calls the actual action (`bundle`, `separate` or `encode`).

In `didjvu.didjvu import Main`, the first `didjvu` is the outer package (which you can install using *pip* for example and corresponding to the `didjvu` directory). The second `didjvu` is the `didjvu.py` file in this directory, `Main` is the object to import, in this case the `Main` class.

Please note that while I am glad to assist anyone with getting started, I cannot and will not explain all code in detail. Most of the architectural decisions do not originate from me (and I might prefer another approach) as well.

**Reply by jsbien** [reply]
I fully understand your position. Thank you very much for your patience!
